File: py/src/imbrace/resources/templates.py
from typing import Any, Dict, List, Optional
from typing_extensions import TypedDict
from ..http import HttpTransport, AsyncHttpTransport


# ── Input shapes ──────────────────────────────────────────────────────────────


class UseCaseInput(TypedDict, total=False):
    """The ``usecase`` half of the create-custom-template payload.

    All fields optional except ``title``. ``agent_type`` distinguishes use case
    type — for Document AI agents pass ``"document_ai"``.
    """
    title: str
    short_description: Optional[str]
    demo_url: Optional[str]
    agent_type: Optional[str]


class DocumentAIAssistantConfig(TypedDict, total=False):
    """Nested config inside ``assistant.document_ai`` for Document AI agents.

    Mirrors the body sent by the iMBRACE webapp when creating a Document AI agent.
    """
    vlm_provider_id: str
    vlm_model: str
    source_languages: List[str]
    handwriting_support: bool
    board_id: str
    time_offset: Optional[str]
    continue_on_failure: bool
    retry_time: int


class AssistantInput(TypedDict, total=False):
    """The ``assistant`` half of the create-custom-template payload.

    For Document AI agents, set ``agent_type="document_ai"`` and populate the
    nested ``document_ai`` config (linking the schema board, VLM provider, etc.).
    """
    name: str
    description: Optional[str]
    mode: Optional[str]                    # "advanced" | "standard"
    model_id: Optional[str]
    provider_id: Optional[str]
    core_task: Optional[str]               # = instructions for the agent
    agent_type: Optional[str]              # "document_ai" | "agent" | ...
    channel: Optional[str]                 # "web" | ...
    temperature: Optional[float]
    workflow_function_call: Optional[List[str]]
    workflow_name: Optional[str]
    credential_name: Optional[str]
    board_ids: Optional[List[str]]
    version: Optional[int]
    document_ai: Optional[DocumentAIAssistantConfig]
    metadata: Optional[Dict[str, Any]]


# ── Response shapes ───────────────────────────────────────────────────────────


class UseCase(TypedDict, total=False):
    _id: str
    public_id: Optional[str]
    doc_name: str                          # "UseCase"
    title: str
    short_description: Optional[str]
    type: Optional[str]                    # "custom" | ...
    agent_type: Optional[str]
    organization_id: Optional[str]
    user_id: Optional[str]
    channel_id: Optional[str]
    assistant_id: Optional[str]
    demo_url: Optional[str]
    version: Optional[int]
    is_deleted: Optional[bool]
    features: Optional[List[Any]]
    tags: Optional[List[Any]]
    suggestion_prompts: Optional[List[Any]]
    how_it_works: Optional[List[Any]]
    supported_channels: Optional[List[Any]]
    integrations: Optional[List[Any]]
    created_at: Optional[str]
    updated_at: Optional[str]
    createdAt: Optional[str]
    updatedAt: Optional[str]


class UseCaseListResponse(TypedDict, total=False):
    data: List[UseCase]


class CreateCustomTemplateResponse(TypedDict, total=False):
    data: UseCase


# ── Resource ──────────────────────────────────────────────────────────────────


class TemplatesResource:
    """Use Case Templates — Sync.

    Wraps ``/v2/backend/templates`` endpoints. The most important method is
    :meth:`create_custom`, which atomically creates a UseCase + Assistant pair
    used by Document AI flows (see also :class:`DocumentAIResource`).

    @param http  HTTP transport
    @param base  Templates base URL (typically ``{gateway}/v2/backend/templates``)

    Example::

        # Document AI: create UseCase + Assistant linked to a schema board
        res = client.templates.create_custom(
            usecase={
                "title": "Receipt Extractor",
                "short_description": "Extract invoice fields",
                "demo_url": "https://chat-widget.imbrace.co",
                "agent_type": "document_ai",
            },
            assistant={
                "name": "Receipt Extractor",
                "mode": "advanced",
                "model_id": "qwen3.5:27b",
                "provider_id": "8cc8769a-...",
                "core_task": "Step 1: Extract Data...",
                "agent_type": "document_ai",
                "channel": "web",
                "temperature": 0.1,
                "version": 2,
                "document_ai": {
                    "vlm_provider_id": "8cc8769a-...",
                    "vlm_model": "qwen3.5:27b",
                    "source_languages": ["English"],
                    "handwriting_support": True,
                    "board_id": "brd_xxx",
                    "continue_on_failure": False,
                    "retry_time": 2,
                },
            },
        )
        usecase_id    = res["data"]["_id"]
        assistant_id  = res["data"]["assistant_id"]
        channel_id    = res["data"]["channel_id"]
    """

    def __init__(self, http: HttpTransport, base: str):
        self._http = http
        self._base = base.rstrip("/")

    def list(self) -> UseCaseListResponse:
        """List use case templates — ``GET /v2/backend/templates``."""
        return self._http.request("GET", self._base).json()

    def create_custom(
        self,
        usecase: UseCaseInput,
        assistant: AssistantInput,
    ) -> CreateCustomTemplateResponse:
        """Create a custom UseCase + Assistant in one POST.

        Routes to ``POST /v2/backend/templates/v2/custom``. Backend auto-creates
        the linked channel, workflow (ActivePieces), and assistant_app, and
        returns the assembled use case (with ``assistant_id``, ``channel_id``).
        """
        body = {"usecase": usecase, "assistant": assistant}
        return self._http.request("POST", f"{self._base}/v2/custom", json=body).json()

    def update_custom(self, usecase_id: str, body: Dict[str, Any]) -> CreateCustomTemplateResponse:
        """Partial update — ``PATCH /v2/backend/templates/{id}/custom``."""
        return self._http.request("PATCH", f"{self._base}/{usecase_id}/custom", json=body).json()

    def delete(self, usecase_id: str) -> None:
        """Delete v1 template — ``DELETE /v2/backend/templates/{id}``."""
        self._http.request("DELETE", f"{self._base}/{usecase_id}")

    def delete_v2(self, usecase_id: str) -> None:
        """Delete v2 template (atomic) — ``DELETE /v2/backend/templates/v2/{id}``.

        Cascades through linked Assistant + Channel. Use this for templates
        created via :meth:`create_custom`.
        """
        self._http.request("DELETE", f"{self._base}/v2/{usecase_id}")


class AsyncTemplatesResource:
    """Use Case Templates — Async. See :class:`TemplatesResource` for full docs."""

    def __init__(self, http: AsyncHttpTransport, base: str):
        self._http = http
        self._base = base.rstrip("/")

    async def list(self) -> UseCaseListResponse:
        res = await self._http.request("GET", self._base)
        return res.json()

    async def create_custom(
        self,
        usecase: UseCaseInput,
        assistant: AssistantInput,
    ) -> CreateCustomTemplateResponse:
        body = {"usecase": usecase, "assistant": assistant}
        res = await self._http.request("POST", f"{self._base}/v2/custom", json=body)
        return res.json()

    async def update_custom(self, usecase_id: str, body: Dict[str, Any]) -> CreateCustomTemplateResponse:
        res = await self._http.request("PATCH", f"{self._base}/{usecase_id}/custom", json=body)
        return res.json()

    async def delete(self, usecase_id: str) -> None:
        await self._http.request("DELETE", f"{self._base}/{usecase_id}")

    async def delete_v2(self, usecase_id: str) -> None:
        await self._http.request("DELETE", f"{self._base}/v2/{usecase_id}")

File: ts/src/resources/templates.ts
import { HttpTransport } from "../http.js"

// ── Input shapes ──────────────────────────────────────────────────────────────

/**
 * The `usecase` half of the create-custom-template payload.
 * `agent_type` distinguishes use case type — for Document AI agents pass `"document_ai"`.
 */
export interface UseCaseInput {
  title: string
  short_description?: string
  demo_url?: string
  agent_type?: string
  [key: string]: unknown
}

/**
 * Nested config inside `assistant.document_ai` for Document AI agents.
 * Mirrors the body sent by the iMBRACE webapp when creating a Document AI agent.
 */
export interface DocumentAIAssistantConfig {
  vlm_provider_id: string
  vlm_model: string
  source_languages: string[]
  handwriting_support: boolean
  board_id: string
  time_offset?: string
  continue_on_failure?: boolean
  retry_time?: number
  [key: string]: unknown
}

/**
 * The `assistant` half of the create-custom-template payload.
 *
 * For Document AI agents: set `agent_type: "document_ai"` and populate the
 * nested `document_ai` config (linking the schema board, VLM provider, etc.).
 */
export interface AssistantInput {
  name: string
  description?: string
  mode?: "advanced" | "standard"
  model_id?: string
  provider_id?: string
  /** = the agent's instructions / system prompt. */
  core_task?: string
  agent_type?: string
  channel?: string
  temperature?: number
  workflow_function_call?: string[]
  workflow_name?: string
  credential_name?: string
  board_ids?: string[]
  version?: number
  document_ai?: DocumentAIAssistantConfig
  metadata?: Record<string, unknown>
  [key: string]: unknown
}

// ── Response shapes ───────────────────────────────────────────────────────────

export interface UseCase {
  _id: string
  public_id?: string
  doc_name?: string
  title: string
  short_description?: string
  type?: string
  agent_type?: string
  organization_id?: string
  user_id?: string
  channel_id?: string
  assistant_id?: string
  demo_url?: string
  version?: number
  is_deleted?: boolean
  features?: unknown[]
  tags?: unknown[]
  suggestion_prompts?: unknown[]
  how_it_works?: unknown[]
  supported_channels?: unknown[]
  integrations?: unknown[]
  created_at?: string
  updated_at?: string
  createdAt?: string
  updatedAt?: string
  [key: string]: unknown
}

export interface UseCaseListResponse {
  data: UseCase[]
}

export interface CreateCustomTemplateResponse {
  data: UseCase
}

// ── Resource ──────────────────────────────────────────────────────────────────

/**
 * Use Case Templates resource.
 *
 * Wraps `/v2/backend/templates` endpoints. The most important method is
 * {@link createCustom}, which atomically creates a UseCase + Assistant pair
 * used by Document AI flows.
 *
 * @example Create a Document AI agent (UseCase + Assistant linked to a schema board)
 * ```ts
 * const res = await client.templates.createCustom({
 *   usecase: {
 *     title: "Receipt Extractor",
 *     short_description: "Extract invoice fields",
 *     demo_url: "https://chat-widget.imbrace.co",
 *     agent_type: "document_ai",
 *   },
 *   assistant: {
 *     name: "Receipt Extractor",
 *     mode: "advanced",
 *     model_id: "qwen3.5:27b",
 *     provider_id: "8cc8769a-...",
 *     core_task: "Step 1: Extract Data...",
 *     agent_type: "document_ai",
 *     channel: "web",
 *     temperature: 0.1,
 *     version: 2,
 *     document_ai: {
 *       vlm_provider_id: "8cc8769a-...",
 *       vlm_model: "qwen3.5:27b",
 *       source_languages: ["English"],
 *       handwriting_support: true,
 *       board_id: "brd_xxx",
 *       continue_on_failure: false,
 *       retry_time: 2,
 *     },
 *   },
 * })
 * const { _id: usecaseId, assistant_id, channel_id } = res.data
 * ```
 */
export class TemplatesResource {
  /**
   * @param http  HTTP transport
   * @param base  Templates base URL (typically `${gateway}/v2/backend/templates`)
   */
  constructor(
    private readonly http: HttpTransport,
    private readonly base: string,
  ) {}

  private get root() {
    return this.base.replace(/\/$/, "")
  }

  /** List use case templates — `GET /v2/backend/templates`. */
  async list(): Promise<UseCaseListResponse> {
    return this.http.getFetch()(this.root, { method: "GET" }).then(r => r.json())
  }

  /**
   * Create a custom UseCase + Assistant in one POST.
   *
   * Routes to `POST /v2/backend/templates/v2/custom`. Backend auto-creates the
   * linked channel, workflow (ActivePieces), and assistant_app, and returns the
   * assembled use case (with `assistant_id`, `channel_id`).
   */
  async createCustom(input: {
    usecase: UseCaseInput
    assistant: AssistantInput
  }): Promise<CreateCustomTemplateResponse> {
    return this.http.getFetch()(`${this.root}/v2/custom`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(input),
    }).then(r => r.json())
  }
}
